### PR TITLE
Add class LayoutLMv2ForRelationExtraction

### PR DESCRIPTION
**Pull request by yuan-wenhua**
# What does this PR do?
Add class LayoutLMv2ForRelationExtraction which in https://github.com/microsoft/unilm/tree/master/layoutlmft

### CODE REVIEW

File: src/transformers/models/layoutlmv2/re.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    def reset_parameters(self):
+        self.bilinear.reset_parameters()
+        self.linear.reset_parameters()
+
+
+class REDecoder(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.entity_emb = nn.Embedding(3, config.hidden_size, scale_grad_by_freq=True)
```
Comment (nlapidot3) [inline src/transformers/models/layoutlmv2/re.py:59]: Sorry for butting in, but do you know of any reason to hard code 3 for the `num_embeddings` parameter (aside from it being the value in `unilm`)? I came across this PR in prepping to implement a version of this same model for a relation extraction task--so thank you for working on this in the first place! All of that to say, I'm asking for selfish reasons 😅

Comment (nlapidot3) [reply src/transformers/models/layoutlmv2/re.py:59]: fwiw I'm guessing the LayoutXLM authors used 3 because they were dealing with forms that only had 3 semantic entity classes (headers, keys, and values).

File: src/transformers/__init__.py [hunk 2]
```diff
@@ -22,8 +22,7 @@
 # to defer the actual importing for when the objects are requested. This way `import transformers` provides the names
 # in the namespace without actually importing anything (and especially none of the backends).
 
-__version__ = "4.16.0.dev0"
-
+__version__ = "4.16.0"
```
Comment (NielsRogge) [inline src/transformers/__init__.py:25]: Don't think you need to change this line.

File: src/transformers/modeling_outputs.py [hunk 3]
```diff
@@ -812,3 +812,41 @@ class Seq2SeqQuestionAnsweringModelOutput(ModelOutput):
     encoder_last_hidden_state: Optional[torch.FloatTensor] = None
     encoder_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
     encoder_attentions: Optional[Tuple[torch.FloatTensor]] = None
+
+@dataclass
+class ReOutput(ModelOutput):
+    """
+    Base class for outputs of relation extraction models.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:819]:
```suggestion
class LayoutLMv2RelationExtractionOutput(ModelOutput):
    """
    Class for outputs of [`LayoutLMv2ForRelationExtraction`].
```

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 4]
```diff
@@ -36,10 +36,12 @@
     QuestionAnsweringModelOutput,
     SequenceClassifierOutput,
     TokenClassifierOutput,
+    ReOutput,
```
Comment (NielsRogge) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:39]: This model output class can be defined within the modeling file itself.

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 5]
```diff
@@ -36,10 +36,12 @@
...
 )
 from ...modeling_utils import PreTrainedModel, apply_chunking_to_forward
 from ...utils import logging
 from .configuration_layoutlmv2 import LayoutLMv2Config
+from .re import REDecoder
```
Comment (NielsRogge) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:44]: You can include the decoder inside `modeling_layoutlmv2.py`.

Our philosophy is a single model = a single script.

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 6]
```diff
@@ -1226,6 +1228,86 @@ def forward(
         )
 
 
+class LayoutLMv2ForRelationExtraction(LayoutLMv2PreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+        self.layoutlmv2 = LayoutLMv2Model(config)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+        self.extractor = REDecoder(config)
+        # Initialize weights and apply final processing
```
Comment (NielsRogge) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1237]:
```suggestion
        
        # Initialize weights and apply final processing
```

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 7]
```diff
@@ -1226,6 +1228,86 @@ def forward(
...
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.layoutlmv2.embeddings.word_embeddings
+
+    @add_start_docstrings_to_model_forward(LAYOUTLMV2_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
+    @replace_return_docstrings(output_type=ReOutput, config_class=_CONFIG_FOR_DOC)
```
Comment (NielsRogge) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1244]:
```suggestion
    @replace_return_docstrings(output_type=LayoutLMv2RelationExtractionOutput, config_class=_CONFIG_FOR_DOC)
```

File: src/transformers/models/layoutlmv2/re.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/layoutlmv2/re.py:59]: Yes in that case, you can replace it with `config.num_labels`.

File: src/transformers/models/layoutlmv2/re.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
+import copy
+
+import torch
+from torch import nn
+from torch.nn import CrossEntropyLoss
+
+
+class BiaffineAttention(torch.nn.Module):
```
Comment (NielsRogge) [inline src/transformers/models/layoutlmv2/re.py:8]:
```suggestion
class LayoutLMv2BiaffineAttention(torch.nn.Module):
```
We usually add a model-specific prefix to each class.

**Review by NielsRogge — COMMENTED** [review]
Thanks for working on this! A few comments:

- the model should be added to the tests (tests/test_modeling_layoutlmv2.py). Normally, if you run `make fixup` locally, it will automatically complain about the fact that this currently isn't the case.
- we do have the philosophy of "single model, single file". Hence, you can include the model specific output class, as well as the relation extraction decoder inside `modeling_layoutlm_v2`.py.

**Comment by quasimik** [comment]
I've [implemented](https://github.com/woflowinc/transformers/tree/add-layoutlmv2-re) the requested changes, but I'm not sure how to contribute to this PR from here.

File: src/transformers/models/layoutlmv2/re.py (hunk 8 above)
Comment (kmurat) [reply src/transformers/models/layoutlmv2/re.py:8]: @NielsRogge 
`self.rel_classifier = BiaffineAttention(config.hidden_size // 2, 2)`
Can you explain why we pass those parameters to BiaffineAttention class?
2 because we it should classify as [(QUESTION, ANSWER), None]?

Comment (kmurat) [reply src/transformers/models/layoutlmv2/re.py:8]: Let's say that we have 5 labels [other, L1, L2, L3, L4] and we don't have a predefined relation for them like [Question, Answer] in this case. Does it mean that I need to pass a number of all the possible combinations (4 * 3 = 12) instead of 2?

**Comment by NielsRogge** [comment]
Hi @quasimik,

could you open up a clean PR such that we can add it?
